src/test/ui/parser/increment.stderr

```diff
@@ -0,0 +1,23 @@
+error: expected expression, found `+`
+  --> $DIR/increment.rs:3:7
+   |
+LL |     i++;
+   |       ^ expected expression
+   |
+   = note: Rust has no dedicated increment and decrement operators
+help: try using `+= 1` instead
+   |
+LL |     i += 1;
+   |       ^^^^
```

Comment:
One issue with using a structured suggestion is that we will suggest invalid code in the case of

```rust
fn main() {
    let mut i = 0;
    while i++ < 5 { /* ... */ }
```

We suggest

```rust
fn main() {
    let mut i = 0;
    while i += 1 < 5 { /* ... */ }
}
```

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0308]: mismatched types
 --> src/main.rs:3:11
  |
3 |     while i += 1 < 5 { /* ... */ }
  |           ^^^^^^^^^^ expected `bool`, found `()`

error[E0277]: cannot add-assign `bool` to `{integer}`
 --> src/main.rs:3:13
  |
3 |     while i += 1 < 5 { /* ... */ }
  |             ^^ no implementation for `{integer} += bool`
  |
  = help: the trait `AddAssign<bool>` is not implemented for `{integer}`

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `playground`

To learn more, run the command again with --verbose.

```

And adding parentheses doesn't fix it:

```rust
fn main() {
    let mut i = 0;
    while (i += 1) < 5 { /* ... */ }
}
```

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0308]: mismatched types
 --> src/main.rs:3:22
  |
3 |     while (i += 1) < 5 { /* ... */ }
  |                      ^ expected `()`, found integer

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground`

To learn more, run the command again with --verbose.

```

-----

So maybe it would be better to just skip the structured suggestion?

Reply:
I'm thinking whether it would make sense to try to suggest `while { i += 1; i } < 5`?

Reply:
That's interesting! The advantage of that is that it will preserve the behavior of `i++` in other languages. Would we always suggest `{ i += 1; i }` or would we sometimes suggest just `i += 1`? I think it would probably be pretty hard to have conditional logic in this case, so I vote for unconditionally using `{ i += 1; i }`, unless there's a simple way to encode the logic.

Also, what should we suggest in the `++i` case? In that case, `{ i += 1; i }` is **not** the correct behavior; rather, the equivalent in Rust is `{ let temp = i; i += 1; temp }`. The problem with suggesting that is that the user may have code like:

```rust
++temp;
```

and then we would suggest

```rust
{ let temp = temp; temp += 1; temp }
```

which is incorrect.

Reply:
So, thinking about it more, maybe simply suggesting `i += 1` in all cases is good enough? Or, alternatively, would it be better to just include a `note` and skip the structured suggestion?

Reply:
You wrote that backwards: preincrement `++i` is `{ i += 1; i }`, and postincrement `i++` is `{ let temp = i; i += 1; temp }`.